common/types.go

```diff
@@ -234,16 +237,302 @@ func init() {
 }
 
 // Address represents the 20 byte address of an Ethereum account.
-type Address [AddressLength]byte
+type InternalAddress [AddressLength]byte
+type ExternalAddress [AddressLength]byte
+
+type Address struct {
+	inner AddressData
+}
+
+type AddressData interface {
+	Bytes() []byte
+	Hash() Hash
+	Hex() string
+	String() string
+	checksumHex() []byte
+	Format(s fmt.State, c rune)
+	SetBytes(b []byte)
+	MarshalText() ([]byte, error)
+	UnmarshalText(input []byte) error
+	UnmarshalJSON(input []byte) error
+	Scan(src interface{}) error
+	Value() (driver.Value, error)
+	Location() *Location
+	IsInChainScope() bool
+}
+
+func NewAddress() Address {
+	var a Address
+	var i InternalAddress
+	a.inner = &i
+	return a
+}
+
+// IsInChainScope checks if an address is a valid account in our node's sharded address space
+func (a Address) IsInChainScope() bool {
+	if a.inner == nil || a.inner.Hash() == ZeroAddr.Hash() {
+		return true
+	}
+	return NodeLocation.ContainsAddress(a)
+}
+
+func (a *Address) SetZero() *InternalAddress {
+	inner := &InternalAddress{0x0000000000000000000000000000000000000000}
+	a.inner = inner
+	return inner
+}
+
+func (a Address) InternalAddress() (*InternalAddress, error) {
+	if a.inner == nil {
+		return &InternalAddress{}, nil
+	}
+	internal, ok := a.inner.(*InternalAddress)
+	if !ok {
+		return nil, ErrInvalidScope
+	}
+	return internal, nil
+}
+
+func (a Address) Equals(b Address) bool {
+	if a.inner == nil && b.inner == nil {
+		return true
+	} else if a.inner == nil || b.inner == nil {
+		return false
+	}
+	return a.Hash() == b.Hash()
+}
 
 // BytesToAddress returns Address with value b.
 // If b is larger than len(h), b will be cropped from the left.
 func BytesToAddress(b []byte) Address {
+	if IsInChainScope(b) {
+		var i InternalAddress
+		i.SetBytes(b)
+		return Address{&i}
+	} else {
+		var e ExternalAddress
+		e.SetBytes(b)
+		return Address{&e}
+	}
+}
+
+func Bytes20ToAddress(b [20]byte) Address {
```

Comment:
I see three constructors:
1) NewAddress()
1) BytesToAddress(b []byte)
1) Bytes20ToAddress(b [20]byte)

`NewAddress()` is only used in one place, and its used just to produce an empty address in case of error. We really shouldn't "construct" an invalid address type. I can think of a few other ways that one error condition could be handled.

`BytesToAddress(b []byte)` provides a more flexible constructor (used by most of the codebase), but doesn't throw an error if too few or too many bytes are supplied. Basically, it is just a less strict version of `Bytes20ToAddress()`. This concerns me, because if we ever have a bug where we don't supply a full 20 bytes, we could enable address collision attacks. E.g. if somewhere in our code is accidentally only using the first N bytes when creating an address, then any address with that same first N bytes would pass through the operation equally... which could yield potentially serious bugs.

`Bytes20ToAddress(b [20]byte)` is only used in a few places, each of which could just have easily used `BytesToAddress()`

**tl;dr:**
I suggest we collapse all of these into one constructor: `NewAddress(b [20]bytes)`. This avoids confusion, and forces users to supply exactly 20 bytes, instead of loosy goosy accepting any number of bytes.

Reply:
`BytesToAddress(b []byte)` already exists in our codebase and also exists in go-ethereum. https://github.com/dominant-strategies/go-quai/blob/main/common/types.go#L241
This is production code in geth, so if there was a bug related to supplying too little bytes, this bug would have already been found, no?
Also, I have removed `NewAddress`

Reply:
Ah, ok, I'm fine with that :+1: